File: typescript/src/bigram.ts
/**
 * Bigram Collocation Model for Thai Word Tokenization in TypeScript.
 */

export class BigramModel {
  private bigrams: Map<string, number> = new Map();
  public alpha: number;

  constructor(alpha: number = 0.15) {
    this.alpha = alpha;
  }

  get size(): number {
    return this.bigrams.size;
  }

  /**
   * Add a bigram transition with frequency count.
   */
  add(prevWord: string, nextWord: string, count: number = 1.0): void {
    this.bigrams.set(`${prevWord}\t${nextWord}`, count);
  }

  /**
   * Calculate transition bonus for a word pair.
   * Higher frequency yields higher cost deduction.
   */
  getBonus(prevWord: string, nextWord: string, wordLen: number = 1): number {
    if (this.alpha <= 0.0 || !prevWord || !nextWord) {
      return 0.0;
    }

    const count = this.bigrams.get(`${prevWord}\t${nextWord}`);
    if (count === undefined) {
      return 0.0;
    }

    const denom = Math.max(1, wordLen);
    return (this.alpha * Math.log(1.0 + count)) / denom;
  }

  /**
   * Load bigram transitions from a TSV string (prev<TAB>next<TAB>count).
   */
  static fromTsv(content: string, alpha: number = 0.15): BigramModel {
    const model = new BigramModel(alpha);
    const lines = content.split(/\r?\n/);
    for (const line of lines) {
      const trimmed = line.trim();
      if (!trimmed || trimmed.startsWith('#')) continue;
      const parts = trimmed.split('\t');
      if (parts.length < 3) continue;
      const [w1, w2, countStr] = parts;
      const count = parseFloat(countStr);
      if (!isNaN(count) && count > 0) {
        model.add(w1, w2, count);
      }
    }
    return model;
  }
}

File: go/bigram.go
package thaibreak

import (
	"bufio"
	"io"
	"math"
	"os"
	"strconv"
	"strings"
	"sync"
)

// BigramModel stores bigram collocation transition counts.
type BigramModel struct {
	mu      sync.RWMutex
	bigrams map[string]float64
	alpha   float64
}

// NewBigramModel creates an empty BigramModel with default alpha = 0.15.
func NewBigramModel() *BigramModel {
	return &BigramModel{
		bigrams: make(map[string]float64),
		alpha:   0.15,
	}
}

// SetAlpha sets the scaling weight of bigram bonuses.
func (m *BigramModel) SetAlpha(alpha float64) {
	m.mu.Lock()
	defer m.mu.Unlock()
	m.alpha = alpha
}

// LoadBigramsTsv loads bigram transitions from a TSV reader (prev \t next \t count).
func LoadBigramsTsv(r io.Reader) (*BigramModel, error) {
	model := NewBigramModel()
	scanner := bufio.NewScanner(r)
	for scanner.Scan() {
		line := strings.TrimSpace(scanner.Text())
		if line == "" || strings.HasPrefix(line, "#") {
			continue
		}
		parts := strings.Split(line, "\t")
		if len(parts) < 3 {
			continue
		}
		w1, w2 := parts[0], parts[1]
		count, err := strconv.ParseFloat(parts[2], 64)
		if err == nil && count > 0 {
			model.Add(w1, w2, count)
		}
	}
	return model, scanner.Err()
}

// LoadBigramsFile loads bigrams from a TSV file path.
func LoadBigramsFile(path string) (*BigramModel, error) {
	f, err := os.Open(path)
	if err != nil {
		return nil, err
	}
	defer f.Close()
	return LoadBigramsTsv(f)
}

// Add inserts or updates a transition count.
func (m *BigramModel) Add(w1, w2 string, count float64) {
	m.mu.Lock()
	defer m.mu.Unlock()
	m.bigrams[w1+"\t"+w2] = count
}

// GetBonus calculates transition bonus for a word pair.
// Higher frequency yields higher cost deduction.
func (m *BigramModel) GetBonus(prevWord, nextWord string, wordLen int) float64 {
	if m == nil || prevWord == "" || nextWord == "" {
		return 0.0
	}
	m.mu.RLock()
	defer m.mu.RUnlock()

	if m.alpha <= 0.0 {
		return 0.0
	}

	count, exists := m.bigrams[prevWord+"\t"+nextWord]
	if !exists {
		return 0.0
	}

	denom := float64(wordLen)
	if denom < 1.0 {
		denom = 1.0
	}
	return (m.alpha * math.Log(1.0+count)) / denom
}
